### PR TITLE
Cell list implementation

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
SUMMARY: Performance "Implement cell list."

#### Purpose of change
When dealing with the many monster-vs-many-monster scenario, we currently have every monster walk the list of every opposing monster to determine which one to target.  To put it mildly, this does not scale (O(M*N) or more conventionally O(N^2)).
In order to reduce the overhead during this scenario, I've come up with adopting a Cell List, which is a data structure for scanning a neighborhood for nearby entities to interact with.

#### Describe the solution
The cell list maintains a 2 or 3 dimensional array of lists, the size of a single cell is substantially smaller than the size of the entire area being considered.
e.g. during monster v monster setup, we would walk the list of monsters once and place each one in it's cell like list[monster.pos().x / CELL_SIZE][monster.pos().y / CELL_SIZE].push_back(monster);
Things get a little more complicated once it's time to iterate over the list.  Starting from our origin (nominally the current monster), we want to visit each other potential enemy in turn in order of increasing distance, and stop as soon as a valid target is encountered.
In order to accomplish this, we produce an iterator that accumulates a list of ordered entities that have been discovered, and populate it as needed when we run out of already-processed entities.

#### Describe alternatives you've considered
Some kind of quadtree or similar, but it turns out these don't actually handle this particular problem all that well unless you have a very highly optimised version of one, and if you don't we actually risk increasing our overhead in the worst case instead of reducing it.

#### Progress
- [x] Basic cell list
- [x] Working with a somewhat representative use case
- [x] Edge case checking
- [x] Handle empty map case

### CODE REVIEW

File: src/cell_list.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY>
+class cell_list_iterator
+{
+    public:
+        // The default constructor creates an end iterator.
+        cell_list_iterator() : range( nullptr ) {};
+        cell_list_iterator( const cell_list_range< T, CELL_SIZE, MAXX, MAXY > *range ) : range( range ) {
+            for( int i = 0; i < range->distance; ++i ) {
+                sorted_elements.push_back( std::vector<std::pair<point, T * >>() );
```
Comment (kevingranade) [inline src/cell_list.h]: This isn't done, and right here is the reason, this is throwing a bad_alloc exception and I'm at my wits end because I have no idea why.

**Comment by TechyBen** [comment]
IIRC Rimworld does similar (for pathing and other things) and even shows the cells in the debug options. So if you want any inspiration, check it out (though no idea if the implementation is of any help or not here).

https://youtu.be/RMBQn_sg7DA

File: src/cell_list.h [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        void add( const point &p, T &value ) {
+            std::pair<point, T *> val( p, &value );
+            cells[p.y / CELL_SIZE][p.x / CELL_SIZE].push_back( val );
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
+            return cell_list_range<T, CELL_SIZE, MAXX, MAXY>( *this, p, distance );
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        std::vector<std::pair<point, T *>> cells[MAXY / CELL_SIZE][MAXX / CELL_SIZE];
```
Comment (jbytheway) [inline src/cell_list.h]: I think these divisions need to be `divide_round_up`, otherwise you're going to end up accessing outside the array.

File: src/cell_list.h [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                    if( distance > range->distance ) {
+                        continue;
+                    }
+                    sorted_elements[distance].push_back( candidate );
+                }
+            } else {
+                // Draw a box...
+                for( int x = min_cell_x; x < max_cell_x; ++x ) {
+                    for( int y = min_cell_y; y < max_cell_y; ++y ) {
+                        for( const std::pair<point, T *> &candidate : range->list->cells[min_cell_y][min_cell_x] ) {
```
Comment (jbytheway) [inline src/cell_list.h]:
```suggestion
                        for( const std::pair<point, T *> &candidate : range->list->cells[y][x] ) {
```
I'm guessing this is what was intended?

Comment (kevingranade) [reply src/cell_list.h]: This whole chunk is a placeholder, so yea that's closer, but I just put it all up for review because I was stuck on a dead end with the vector initialization thing, which turned out to be a red herring.

**Comment by kevingranade** [comment]
This is about 3/4 baked or so, I need some more unit tests because this thing is built almost entirely of edge cases.

File: src/cell_list.h [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                    sorted_elements[distance].push_back( candidate );
+                }
+            } else {
+                int min_cell_x = range->origin_cell.x - cell_ranks_loaded;
+                int min_cell_y = range->origin_cell.y - cell_ranks_loaded;
+                int max_cell_x = range->origin_cell.x + cell_ranks_loaded;
+                int max_cell_y = range->origin_cell.y + cell_ranks_loaded;
+                bool min_x_inbounds = min_cell_x > 0;
+                bool min_y_inbounds = min_cell_y > 0;
+                bool max_x_inbounds = max_cell_x <= MAXX / CELL_SIZE;
```
Comment (jbytheway) [inline src/cell_list.h]: If you do switch to `divide_round_up` then all these uses need to change too.  Could make a `static constexpr point cell_dimensions()` member function in `cell_list`.

File: src/cell_list.h [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                if( min_x_inbounds ) {
+                    add( range->list->cells[y][min_cell.x] );
+                }
+                if( max_x_inbounds ) {
+                    add( range->list->cells[y][max_cell.x] );
+                }
+            }
+        }
+        const cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
+        std::vector<std::vector<std::pair<point, T>>> sorted_elements;
```
Comment (jbytheway) [inline src/cell_list.h]: Ideally iterators should be cheap to copy.  I'm wondering how much of the data herein you could safely move to the `cell_list_range`.

File: src/cell_list.h [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = range->origin_cell - load_offset;
+            const point max_cell = range->origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < divide_round_up( MAXX, CELL_SIZE );
+            const bool max_y_inbounds = max_cell.y < divide_round_up( MAXY, CELL_SIZE );
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x > MAXX / CELL_SIZE ) {
```
Comment (jbytheway) [inline src/cell_list.h]:
```suggestion
                if( x < 0 || x > divide_round_up( MAXX, CELL_SIZE ) ) {
```
But I think it might be even cleaner to add a `static constexpr point cell_dim()` member to `cell_list` and use that everywhere you need to know the dimensions.

File: src/cell_list.h [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        }
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, range->origin );
+            if( distance > range->distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            const point rank_offset{ CELL_SIZE, CELL_SIZE };
```
Comment (jbytheway) [inline src/cell_list.h]:
```suggestion
            constexpr point rank_offset{ CELL_SIZE, CELL_SIZE };
```

File: src/cell_list.h [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            max_bounds = min_bounds + point( CELL_SIZE - 1, CELL_SIZE - 1 );
+            distance_to_nearest_cell_wall = calc_distance();
+            // Special case for first cell.
+            add( range->list->cells[range->origin_cell.y][range->origin_cell.x] );
+            current = sorted_elements[current_distance].begin();
+            if( sorted_elements[current_distance].empty() ) {
+                ++( *this );
+            }
+        }
+        std::pair<point, T> &operator*() {
```
Comment (jbytheway) [inline src/cell_list.h]:
```suggestion
        std::pair<point, T> &operator*() const {
```

File: src/cell_list.h [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            assert( range );
+            return *current;
+        }
+        std::pair<point, T> *operator->() {
```
Comment (jbytheway) [inline src/cell_list.h]:
```suggestion
        std::pair<point, T> *operator->() const {
```

File: src/cell_list.h (hunk 5 above)
Comment (kevingranade) [reply src/cell_list.h]: I was avoiding that because it seems more straightforward for it to be in the iterator, but now that I think about it all I have to do is shift the data...

File: src/cell_list.h (hunk 6 above)
Comment (kevingranade) [reply src/cell_list.h]: yes, but num_cells_x() and num_cells_y()

File: src/cell_list.h (hunk 5 above)
Comment (kevingranade) [reply src/cell_list.h]: looks like I can move everything but this:
```
        const cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
        int current_distance = 1;
        typename std::vector<std::pair<point, T>>::iterator current;
```

Comment (jbytheway) [reply src/cell_list.h]: Yes, that should be better.  Make sure you have a test where you iterate over the same range twice.

**Comment by jbytheway** [comment]
Looks like ASan has picked up something fishy in the tests.

**Comment by kevingranade** [comment]
I *think* I see the problem, I had an off-by-one in the iterator increment that could cause an access off the end of the main vector.  I think I may have run into the same problem while refactoring the data into the range object that caused a segfault.

File: src/cell_list.h (hunk 5 above)
Comment (kevingranade) [reply src/cell_list.h]: Of course this *looked* simple, but now I have crashing and other weirdness after the refactor, so I'm retreating away from 99% done at great velocity >_<

**Comment by kevingranade** [comment]
One *simple refactor* and about 150 lines changed later, and I think I'm actually done.

File: src/cell_list.h [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+ * It is used to optimize lookup of nearby entities.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list
+{
+    public:
+        cell_list() = default;
+        void add( const point &p, T value ) {
+            std::pair<point, T> val( p, value );
+            cells[p.x / CELL_SIZE][p.y / CELL_SIZE].push_back( val );
```
Comment (codemime) [inline src/cell_list.h]:
```suggestion
            cells[p.x / CELL_SIZE][p.y / CELL_SIZE].push_back( p, value );
```
would be arguably neater [ Dunno how to a make multiline "suggestion" to delete the upper line ].

File: src/cell_list.h [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
```
Comment (codemime) [inline src/cell_list.h]: Can it become `const`?

File: src/cell_list.h [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+ * First it visits the cell the origin is in, determining the distance to the nearest cell wall,
+ * and iterating over every entity closer than that distance. This guarantees that there is no
+ * closer entity than one found via this process. Once that list is exhausted, it loads one row of
+ * cells around the origin cell and repeating the above process.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range
+{
+    public:
+        cell_list_range( const cell_list< T, CELL_SIZE, MAXX, MAXY > &list, const point &p,
```
Comment (codemime) [inline src/cell_list.h]: I'd make the constructor `private` so that only `cell_list` could produce instances.

File: src/cell_list.h [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+                         int distance ) : list( &list ), origin( p ), origin_cell( p / CELL_SIZE ),
+            max_distance( distance ), sorted_elements( distance + 1, std::vector<std::pair<point, T >>() ) {
```
Comment (codemime) [inline src/cell_list.h]: You presumably won't need all `distance + 1` elements in the most common case, but the constructor creates them all anyways. Wouldn't it be better to add them on demand (optionally only `reserve`ing in the constructor)?

File: src/cell_list.h [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = origin_cell - load_offset;
+            const point max_cell = origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < list->num_cells_x();
+            const bool max_y_inbounds = max_cell.y < list->num_cells_x();
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x >= list->num_cells_x() ) {
```
Comment (codemime) [inline src/cell_list.h]: These checks can be brought out of the loop and embodied into `min_cell` / `max_cell` (same for `y`).

File: src/cell_list.h [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+                if( max_x_inbounds ) {
+                    add( list->cells[max_cell.x][y] );
+                }
+            }
+        }
+        const cell_list<T, CELL_SIZE, MAXX, MAXY> *list;
+        const point origin;
+        const point origin_cell; // = origin / CELL_SIZE
+        const int max_distance;
+        std::vector<std::vector<std::pair<point, T>>> sorted_elements;
```
Comment (codemime) [inline src/cell_list.h]: Copying whole pairs from `cell_list` can be somewhat expensive (especially assuming `T` is something non-trivial). Maybe store iterators / pointers instead? However, this implies immutability of `list` during traversal, but I suppose it's a prerequisite anyways.

File: src/cell_list.h [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+            }
+            return current == rhs.current;
+        }
+        bool operator!=( const cell_list_iterator &rhs ) const {
+            return !( *this == rhs );
+        }
+
+    private:
+        cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
+        int current_distance = 1;
```
Comment (codemime) [inline src/cell_list.h]: `sorted_elements` starts with zero, but it's never visited. Is the discrepancy intentional?

File: src/cell_list.h (hunk 10 above)
Comment (kevingranade) [reply src/cell_list.h]: I meant to do that and forgot, thanks.

File: src/cell_list.h (hunk 11 above)
Comment (kevingranade) [reply src/cell_list.h]: Sort of? I could mark sorted_elements and friends mutable, but that's most of the data structure so it doesn't seem worthwhile.

File: src/cell_list.h (hunk 13 above)
Comment (kevingranade) [reply src/cell_list.h]: I mostly just don't want to complicate things by checking for size and doing push_back() into the vector on every insertion into sorted_elements.  Empty vectors are extremely lightweight, I don't want to complicate this unless profiling shows this as a bottleneck.
Also optimization could go either way because a resize could result in copying an existing sorted_elements vector, which could be quite expensive.

File: src/cell_list.h (hunk 14 above)
Comment (kevingranade) [reply src/cell_list.h]: I thought so at first, but I can't move either min_cell or max_cell because that would shift the box and make the behavior incorrect, e.g. if you have values like this
```
(-1, -1), (0, -1), ( 1, -1),
(-1,  0),              ( 1,  0),
(-1,  1), ( 0,  1), ( 1,  1)
```
Then adjusting min_cell and max_cell would result in:
```
( 0,  0), ( 1, 0),
( 0,  1), ( 1, 1)
```
Which is incorrect, what we want is
```
           (1, 0),
(0, 1), (1, 1)
```

File: src/cell_list.h (hunk 16 above)
Comment (kevingranade) [reply src/cell_list.h]: Just making indexing easier, there's already a ton of index and bounds checking going on here without risking one more off by one error to save a single empty vector.

File: src/cell_list.h (hunk 15 above)
Comment (kevingranade) [reply src/cell_list.h]: Yes this can easily capture a pointer instead of make a copy.

**Comment by kevingranade** [comment]
Just to update people who are interested, I'm still working on this.
It "works" as of the most recent commit, but it's not IMO optimal enough to merge as-is, so I'm looking at micro-optimizing it.
For a representative use case in game of waiting above a lab with z-levels enabled (this is a good use case because these locations have many monsters with different faction hostilities, and they are frustrated because they frequently can't see each other), the changes are a wash, the time spent processing targets is halved, but the overhead of operating the cell list overwhelms that.

I'm fairly certain that in a more pathological case the cell list will come out far ahead, but I'm hopeful I can get at least a moderate speedup in the more typical use cases as well.

The other holdup is rebasing these changes on top of #33244, which significantly changed how the faction tracker works.

Finally, I found I need to inject tripoints into the cell list instead of points.

**Comment by kevingranade** [comment]
IDK what's going on here, after a hell of a lot of overhaul to cell list, it's now producing some bizarre results.

The reproduction case is fairly straightforward:
```
    cell_list<int, 3, 10, 10> list;
    list.add( { 2, 2, 0 }, 0 );
    cell_list_range<int, 3, 10, 10> range = list.get( {1, 1, 0}, 1 );
    auto it = range.begin();
    auto end = range.end();
    while( it != end ) {
        const std::pair<tripoint, int> &ref = *it;
        CHECK( ref.first == tripoint( 2, 2, 0 ) );
        CHECK( ref.second == 0 );
        ++it;
    }
```
In a debugger everything is fine until the first CHECK statement, at which point the value of ref changes from ``(2, 2, 0), 0`` to ``(74923864,32767,74923464), 32767``.
I've been banging my head against this for hours, so a second look would be greatly appreciated.

File: src/cell_list.h [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - origin.y );
+            }
+            return distance;
+        }
+        void add( const std::map<tripoint, T> &cells ) {
+            for( const std::pair<tripoint, T> &candidate : cells ) {
```
Comment (BevapDin) [inline src/cell_list.h]: Here `candidate` is a temporary object (the address of it is added to `sorted_elements`). Iterating over a map yields `const std::pair<const tripoint, T>` (note the second const). Otherwise you would be able to iterate over a non-const map and would have write-access to the keys of it.

Or simpler: `decltype( cells )::value_type` (or the like).

Comment (kevingranade) [reply src/cell_list.h]: Are you saying that since the const-ness doesn't match, I'm getting a copy instead of a reference, or that no matter what I do it's a temporary and it's invalid to capture its address?

Comment (kevingranade) [reply src/cell_list.h]: Evidently the former. Thanks a ton, I was nowhere near the solution.

**Comment by kevingranade** [comment]
Current status is this works, but the performance benefits are occasional at best, and in some cases it makes things significantly worse.
In particular there's a pathological (but unfortunately common) scenario where there are a significant number of monsters hostile to each other in a given area, but none of them can see each other.  In this scenario the new code exhaustively searches all enemies in range, causing overheads similar to what the previous code encountered in the same scenario.  The overhead of building and iterating over the cell list roughly equals the overhead of exhaustively iterating over the entire faction list while trivially rejecting enemies who are out of range.  
The particularly good outcome for the new code should be a large scrum of factions with good visibility to nearby enemies so that it can short-circuit out of the search a significant proportion of the time, but I haven't seen evidence of that happening in any of the scenarios I've built.

I'm revisiting other options for optimization of monster->monster vision, and I'll revisit this approach later.

**Comment by kevingranade** [comment]
Shelving this because I can't use it right now.
